base/system/services/rpcserver.c

```diff
@@ -1752,58 +1765,60 @@ RSetServiceStatus(
     /* Restore the previous service type */
     lpService->Status.dwServiceType = dwPreviousType;
 
-    /* Dereference a stopped service */
-    if ((lpServiceStatus->dwServiceType & SERVICE_WIN32) &&
-        (lpServiceStatus->dwCurrentState == SERVICE_STOPPED))
+    DPRINT("Service %S changed state %d to %d\n", lpService->lpServiceName, dwPreviousState, lpServiceStatus->dwCurrentState);
+
+    if (lpServiceStatus->dwCurrentState != SERVICE_STOPPED &&
+        dwPreviousState == SERVICE_STOPPED)
+    {
+        /* Keep a reference on all non stopped services */
+        ScmReferenceService(lpService);
+        DPRINT("Service %S has %d references after starting\n", lpService->lpServiceName, lpService->RefCount);
+    }
+
+    /* Check if the service just stopped */
+    if (lpServiceStatus->dwCurrentState == SERVICE_STOPPED &&
+        dwPreviousState != SERVICE_STOPPED)
     {
-        /* Decrement the image run counter */
-        lpService->lpImage->dwImageRunCount--;
+        DPRINT("Service %S, currentstate: %d, prev: %d\n", lpService->lpServiceName, lpServiceStatus->dwCurrentState, dwPreviousState);
 
-        /* If we just stopped the last running service... */
-        if (lpService->lpImage->dwImageRunCount == 0)
+        /* 
+         * The service just changed its status to stopped.
+         * Create a thread that will complete the stop sequence.
+         * This thread will remove the reference that was added when the service started
+         * This will ensure that the service will remain valid as long as this reference is still held 
+         */
+        hStopThread = CreateThread(NULL,
```

Comment:
This is broken. Since the database is unlocked right after this, lpService could have already been deleted by someone else when the thread runs and it would access an invalid pointer. Why run this in a different thread at all? Why not call it directly while holding the lock?

Reply:
Unless an extra-reference is added before creating the thread (and then the thread should dereference the service, or the caller should do that in case thread creation fails)?

Reply:
That is exactly the reason why i moved service deletion to happen when ref count of a service reaches 0. Keeping a reference to thw handle will also defer deleting it and if it does gets markes for deletion in the meantime, it will get deleted when the thread removes this extra referencw.